server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/get/TransportGetSnapshotsAction.java

```diff
@@ -81,73 +88,107 @@ protected ClusterBlockException checkBlock(GetSnapshotsRequest request, ClusterS
     @Override
     protected void masterOperation(final GetSnapshotsRequest request, final ClusterState state,
                                    final ActionListener<GetSnapshotsResponse> listener) {
-        try {
-            final String repository = request.repository();
-            final Map<String, SnapshotId> allSnapshotIds = new HashMap<>();
-            final List<SnapshotInfo> currentSnapshots = new ArrayList<>();
-            for (SnapshotInfo snapshotInfo : snapshotsService.currentSnapshots(repository)) {
-                SnapshotId snapshotId = snapshotInfo.snapshotId();
-                allSnapshotIds.put(snapshotId.getName(), snapshotId);
-                currentSnapshots.add(snapshotInfo);
-            }
+        final String[] repositories = request.repositories();
+        transportService.sendRequest(transportService.getLocalNode(), GetRepositoriesAction.NAME,
+                new GetRepositoriesRequest(repositories),
+                new ActionListenerResponseHandler<>(
+                        ActionListener.wrap(
+                                response ->
+                                        // switch to GENERIC thread pool because it might be long running operation
+                                        threadPool.executor(ThreadPool.Names.GENERIC).execute(
+                                                () -> getMultipleReposSnapshotInfo(response.repositories(), request.snapshots(),
+                                                        request.ignoreUnavailable(), request.verbose(), listener)),
+                                listener::onFailure),
+                        GetRepositoriesResponse::new));
+    }
 
-            final RepositoryData repositoryData;
-            if (isCurrentSnapshotsOnly(request.snapshots()) == false) {
-                repositoryData = snapshotsService.getRepositoryData(repository);
-                for (SnapshotId snapshotId : repositoryData.getAllSnapshotIds()) {
-                    allSnapshotIds.put(snapshotId.getName(), snapshotId);
-                }
-            } else {
-                repositoryData = null;
+    private void getMultipleReposSnapshotInfo(List<RepositoryMetaData> repos, String[] snapshots, boolean ignoreUnavailable,
+                                              boolean verbose, ActionListener<GetSnapshotsResponse> listener) {
+        final GroupedActionListener<GetSnapshotsResponse.Response> groupedActionListener =
+                new GroupedActionListener<>(
+                        ActionListener.map(listener, responses -> {
+                            assert repos.size() == responses.size();
+                            return new GetSnapshotsResponse(responses);
+                        }), repos.size());
+
+        // run concurrently for all repos on GENERIC thread pool
+        for (final RepositoryMetaData repo : repos) {
+            threadPool.executor(ThreadPool.Names.GENERIC).execute(
+                    () -> {
```

Comment:
I think we should use `ActionRunnable` here. Otherwise, if we ever run an `Exception` that we don't expect we simply loose reference of the `groupedActionListener` quietly.